File: genui/01_foundations/step_02_declarative_tree/partial_json.py
"""Step 02 - a tolerant parser for JSON that is still arriving.

`parse_partial(text)` returns the value the text describes so far. Open
strings are closed, open objects and arrays are closed, a key without a
value is dropped, a number or literal cut in the middle is dropped. Every
container that had to be closed by the parser comes back as a PartialDict
or PartialList, so a renderer can tell "closed by the model" from "closed
by us". The same rules live in page/partial-json.mjs.
"""

import re

NUMBER = re.compile(r"[-+0-9.eE]+")
LITERALS = {"true": True, "false": False, "null": None}


class PartialDict(dict):
    """An object the text had not closed yet."""


class PartialList(list):
    """An array the text had not closed yet."""


def is_partial(value):
    return isinstance(value, (PartialDict, PartialList))


class Incomplete(Exception):
    """End of text inside a value with nothing to show for it yet."""


class Parser:
    def __init__(self, text):
        self.text = text
        self.i = 0

    def at_end(self):
        return self.i >= len(self.text)

    def peek(self):
        return self.text[self.i]

    def skip_ws(self):
        while not self.at_end() and self.peek() in " \t\r\n":
            self.i += 1

    def value(self):
        self.skip_ws()
        if self.at_end():
            raise Incomplete
        c = self.peek()
        if c == "{":
            return self.obj()
        if c == "[":
            return self.arr()
        if c == '"':
            return self.string()[0]
        if c in "tfn":
            return self.literal()
        if c == "-" or c.isdigit():
            return self.number()
        raise ValueError(f"unexpected {c!r} at {self.i}")

    def obj(self):
        self.i += 1  # {
        out = PartialDict()
        while True:
            self.skip_ws()
            if self.at_end():
                return out
            c = self.peek()
            if c == "}":
                self.i += 1
                return dict(out)  # closed by the model: a plain dict
            if c == ",":
                self.i += 1
                continue
            if c != '"':
                raise ValueError(f"expected a key at {self.i}")
            key, closed = self.string()
            if not closed:
                return out  # the key itself was cut
            self.skip_ws()
            if self.at_end():
                return out  # key but no colon yet
            if self.peek() != ":":
                raise ValueError(f"expected ':' at {self.i}")
            self.i += 1
            try:
                out[key] = self.value()
            except Incomplete:
                return out  # key and colon, but no value yet

    def arr(self):
        self.i += 1  # [
        out = PartialList()
        while True:
            self.skip_ws()
            if self.at_end():
                return out
            c = self.peek()
            if c == "]":
                self.i += 1
                return list(out)
            if c == ",":
                self.i += 1
                continue
            try:
                out.append(self.value())
            except Incomplete:
                return out

    def string(self):
        """(text, closed). An open string returns what it has so far."""
        self.i += 1  # opening quote
        parts = []
        while not self.at_end():
            c = self.peek()
            if c == '"':
                self.i += 1
                return "".join(parts), True
            if c == "\\":
                if self.i + 1 >= len(self.text):
                    break  # a lone backslash at the end: wait for the rest
                e = self.text[self.i + 1]
                if e == "u":
                    hex_digits = self.text[self.i + 2 : self.i + 6]
                    if len(hex_digits) < 4:
                        break
                    parts.append(chr(int(hex_digits, 16)))
                    self.i += 6
                    continue
                parts.append({"n": "\n", "t": "\t", "r": "\r", "b": "\b", "f": "\f"}.get(e, e))
                self.i += 2
                continue
            parts.append(c)
            self.i += 1
        self.i = len(self.text)  # a cut escape is consumed too: the rest is coming
        return "".join(parts), False

    def number(self):
        match = NUMBER.match(self.text, self.i)
        raw = match.group(0)
        self.i = match.end()
        if self.at_end() and raw[-1] in "-+.eE":
            raise Incomplete  # "-", "1.", "2e": more digits are coming
        return int(raw) if re.fullmatch(r"-?\d+", raw) else float(raw)

    def literal(self):
        rest = self.text[self.i : self.i + 5]
        for word, value in LITERALS.items():
            if rest.startswith(word):
                self.i += len(word)
                return value
            if word.startswith(rest) and self.i + len(rest) >= len(self.text):
                self.i = len(self.text)
                raise Incomplete  # "tr", "nul": the rest is coming
        raise ValueError(f"bad literal at {self.i}")


def parse_partial(text):
    """The value so far, or None when nothing parseable has arrived."""
    parser = Parser(text)
    parser.skip_ws()
    if parser.at_end():
        return None
    try:
        return parser.value()
    except Incomplete:
        return None

File: genui/01_foundations/step_02_declarative_tree/page/partial-json.mjs
// Step 02 - a tolerant parser for JSON that is still arriving.
// Mirrors partial_json.py: open strings, objects and arrays are closed; a key
// without a value, or a number or literal cut in the middle, is dropped.
// Containers the parser had to close are recorded in PARTIAL, so a renderer
// can tell "closed by the model" from "closed by us". Text that is not JSON
// at all (a code fence, prose) throws: the caller keeps its last render.

export const PARTIAL = new WeakSet();

export function isPartial(value) {
  return typeof value === "object" && value !== null && PARTIAL.has(value);
}

class Incomplete extends Error {}

const LITERALS = { true: true, false: false, null: null };
const ESCAPES = { n: "\n", t: "\t", r: "\r", b: "\b", f: "\f" };
const JSON_NUMBER = /^-?\d+(\.\d+)?([eE][-+]?\d+)?$/;
const HEX4 = /^[0-9a-fA-F]{4}$/;
export const MAX_DEPTH = 64; // deeper than any layout; a runaway stream cannot overflow the stack

class Parser {
  constructor(text) {
    this.text = text;
    this.i = 0;
    this.depth = 0;
  }
  atEnd() { return this.i >= this.text.length; }
  peek() { return this.text[this.i]; }
  skipWs() { while (!this.atEnd() && " \t\r\n".includes(this.peek())) this.i += 1; }

  value() {
    this.skipWs();
    if (this.atEnd()) throw new Incomplete();
    const c = this.peek();
    if (c === "{" || c === "[") {
      if (++this.depth > MAX_DEPTH) throw new Error(`nested deeper than ${MAX_DEPTH} at ${this.i}`);
      try { return c === "{" ? this.obj() : this.arr(); } finally { this.depth -= 1; }
    }
    if (c === '"') return this.string()[0];
    if ("tfn".includes(c)) return this.literal();
    if (c === "-" || (c >= "0" && c <= "9")) return this.number();
    throw new Error(`unexpected ${JSON.stringify(c)} at ${this.i}`);
  }

  obj() {
    this.i += 1; // {
    const out = {};
    PARTIAL.add(out);
    for (;;) {
      this.skipWs();
      if (this.atEnd()) return out;
      const c = this.peek();
      if (c === "}") { this.i += 1; PARTIAL.delete(out); return out; }
      if (c === ",") { this.i += 1; continue; }
      if (c !== '"') throw new Error(`expected a key at ${this.i}`);
      const [key, closed] = this.string();
      if (!closed) return out;
      this.skipWs();
      if (this.atEnd()) return out;
      if (this.peek() !== ":") throw new Error(`expected ':' at ${this.i}`);
      this.i += 1;
      try { out[key] = this.value(); } catch (e) { if (e instanceof Incomplete) return out; throw e; }
    }
  }

  arr() {
    this.i += 1; // [
    const out = [];
    PARTIAL.add(out);
    for (;;) {
      this.skipWs();
      if (this.atEnd()) return out;
      const c = this.peek();
      if (c === "]") { this.i += 1; PARTIAL.delete(out); return out; }
      if (c === ",") { this.i += 1; continue; }
      try { out.push(this.value()); } catch (e) { if (e instanceof Incomplete) return out; throw e; }
    }
  }

  string() {
    // [text, closed]. An open string returns what it has so far.
    this.i += 1;
    let parts = "";
    while (!this.atEnd()) {
      const c = this.peek();
      if (c === '"') { this.i += 1; return [parts, true]; }
      if (c === "\\") {
        if (this.i + 1 >= this.text.length) break;
        const e = this.text[this.i + 1];
        if (e === "u") {
          const hex = this.text.slice(this.i + 2, this.i + 6);
          if (hex.length < 4) break;
          if (!HEX4.test(hex)) throw new Error(`bad unicode escape at ${this.i}`);
          parts += String.fromCharCode(parseInt(hex, 16));
          this.i += 6;
          continue;
        }
        parts += ESCAPES[e] ?? e;
        this.i += 2;
        continue;
      }
      parts += c;
      this.i += 1;
    }
    this.i = this.text.length; // a cut escape is consumed too: the rest is coming
    return [parts, false];
  }

  number() {
    const match = /[-+0-9.eE]+/y;
    match.lastIndex = this.i;
    const raw = match.exec(this.text)[0];
    this.i += raw.length;
    if (/[-+.eE]$/.test(raw) && this.atEnd()) throw new Incomplete(); // "-", "1.", "2e": more digits are coming
    if (!JSON_NUMBER.test(raw)) throw new Error(`bad number at ${this.i}`);
    return Number(raw);
  }

  literal() {
    const rest = this.text.slice(this.i, this.i + 5);
    for (const [word, value] of Object.entries(LITERALS)) {
      if (rest.startsWith(word)) { this.i += word.length; return value; }
      if (word.startsWith(rest) && this.i + rest.length >= this.text.length) {
        this.i = this.text.length;
        throw new Incomplete(); // "tr", "nul": the rest is coming
      }
    }
    throw new Error(`bad literal at ${this.i}`);
  }
}

export function parsePartial(text) {
  // The value so far, or null when nothing parseable has arrived.
  const parser = new Parser(text);
  parser.skipWs();
  if (parser.atEnd()) return null;
  try { return parser.value(); } catch (e) { if (e instanceof Incomplete) return null; throw e; }
}
